docs/language/ql-training-rst/cpp/global-data-flow-cpp.rst

```diff
@@ -0,0 +1,332 @@
+        arg = c.getAnArgument() and
+        arg = pred.asExpr() and
+        arg.getFullyConverted().getUnderlyingType() instanceof PointerType   and
+        arg = lv.getAnAccess() and
+        succ.asUninitialized() = lv
+      )
+    }
+    ...
+  }
+
+Defining sanitizers
+===================
+
+Add a sanitizer, stopping propagation at parameters of formatting functions, to avoid double-reporting:
+
+.. code-block:: ql
+
+  class TaintedFormatConfig extends TaintTracking::Configuration {
+    override predicate isSanitizer(DataFlow::Node nd) {
+      exists (FormattingFunction ff, int idx |
+        idx = ff.getFormatParameterIndex() and
+        nd = DataFlow::parameterNode(ff.getParameter(idx))
+      )
+    }
+    ...
+  }
+
+Data flow models
+================
+
+- To provide models of data/taint flow through library functions, you can implement subclasses of ``DataFlowFunction`` (from ``semmle.code.cpp.models.interfaces.DataFlow``) and ``TaintFunction`` (from ``semmle.code.cpp.models.interfaces.Taint``), respectively
+
+- Example: model of taint flow from third to first parameter of ``memcpy``
+
+   .. code-block:: ql
+   
+     class MemcpyFunction extends TaintFunction {
+           MemcpyFunction() { this.hasName("memcpy") }
+           override predicate hasTaintFlow(FunctionInput i,    FunctionOutput o) 
+               i.isInParameter(2) and o.isOutParameterPointer(0)
+           }
+       }
+   
+.. note::
+
+  See the API documentation for more details about ``FunctionInput`` and ``FunctionOutput``.
+
+.. rst-class:: end-slide
+
+Extra slides
+============
+
+Exercise: How not to do global data flow
+========================================
+
+Implement a flowStep predicate extending localFlowStep with steps through function calls and returns. Why might we not want to use this?
+
+.. code-block:: ql
+
+   predicate stepIn(Call c, DataFlow::Node arg, DataFlow::ParameterNode parm) {
+     exists(int i | arg.asExpr() = c.getArgument(i) |
+       parm.asParameter() = c.getTarget().getParameter(i))
+   }
+   
+   predicate stepOut(Call c, DataFlow::Node ret, DataFlow::Node res) {
+     exists(ReturnStmt retStmt | retStmt.getEnclosingFunction() = c.getTarget() |
+       ret.asExpr() = retStmt.getExpr() and res.asExpr() = c)
+   }
+   
+   predicate flowStep(DataFlow::Node pred, DataFlow::Node succ) {
+     DataFlow::localFlowStep(pred, succ) or
+     stepIn(_, pred, succ) or
+     stepOut(_, pred, succ)
+   }
+
+Mismatched calls and returns
+============================
+
+.. container:: column-left
+
+   .. code-block:: ql
+
+      char *logFormat(char *fmt) {
+        log("Observed format string %s.", fmt);
+        return fmt;
+      }
+      
+      ...
+      char *dangerousFmt = unvalidatedUserData();
+      printf(logFormat(dangerousFmt), args);
+      ...
+      char *safeFmt = "Hello %s!";
+      printf(logFormat(safeFmt), name);
+
+.. container:: column-right
+
+  Infeasible path due to mismatched call/return pair!
```

Comment:
Do we need either some notes, or to replace the code with an image to provide the information from the arrows in the original version of this slide?

Reply:
Yeah, we probably will.